rex/bias_correction.py

```diff
@@ -0,0 +1,578 @@
+        """Clean any kwargs inputs (e.g., dist, relative) that might be
+        provided as an array and must be collapsed into a single string or
+        boolean value"""
+        unique = np.unique(inp)
+        msg = ('_QuantileDeltaMapping kwargs must have only one unique input '
+               'even if being called with arrays as part of reV but found: {}'
+               .format(unique))
+        assert len(unique) == 1, msg
+
+        while isinstance(inp, np.ndarray):
+            inp = inp[0]
+
+        return inp
+
+    def _clean_params(self, params, arr_shape):
+        """Verify and clean 2D parameter arrays for passing into empirical
+        distribution or scipy continuous distribution functions.
+
+        Parameters
+        ----------
+        params : np.ndarray
+            Input params shape should be (space, N) where N is the number of
+            parameters for the distribution.
+        arr_shape : tuple
+            Array shape should be (time, space).
+
+        Returns
+        -------
+        params : np.ndarray | list
+            If a scipy continuous dist is set, this output will be params
+            unpacked along axis=1 into a list so that the list entries
+            represent the scipy distribution parameters
+            (e.g., shape, scale, loc) and each list entry is of shape (space,)
+        """
+
+        msg = f'params must be 2D array but received {type(params)}'
+        assert isinstance(params, np.ndarray), msg
+
+        if len(params.shape) == 1:
+            params = np.expand_dims(params, 0)
+
+        msg = (f'params must be 2D array of shape ({arr_shape[1]}, N) '
+               f'but received shape {params.shape}')
+        assert len(params.shape) == 2, msg
+        assert params.shape[0] == arr_shape[1], msg
+
+        if self.scipy_dist is not None:
+            params = [params[:, i] for i in range(params.shape[1])]
+
+        return params
+
+    def _get_quantiles(self, n_samples):
+        """If dist='empirical', this will get the quantile values for the CDF
+        x-values specified in the input params"""
+
+        if self.sampling == 'linear':
+            quantiles = np.linspace(0, 1, n_samples)
+
+        elif self.sampling == 'log':
+            quantiles = np.logspace(0, 1, n_samples, base=self.log_base)
+            quantiles = (quantiles - 1) / (self.log_base - 1)
+
+        elif self.sampling == 'invlog':
+            quantiles = np.logspace(0, 1, n_samples, base=self.log_base)
+            quantiles = (quantiles - 1) / (self.log_base - 1)
+            quantiles = np.array(sorted(1 - quantiles))
+
+        else:
+            msg = ('sampling option must be linear, log, or invlog, but '
+                   'received: {}'.format(self.sampling))
+            logger.error(msg)
+            raise KeyError(msg)
+
+        return quantiles
+
+    def cdf(self, x, params):
+        """Run the CDF function e.g., convert physical variable to quantile"""
+
```

Comment:
If you're at all interested in decreasing indentation, you could do 

```
if self.scipy_dist is not None:
    return self.scipy_dist.cdf(x, *params)

p = np.zeros_like(x)
for idx in range(x.shape[1]):
    xp = params[idx, :]
    fp = self._get_quantiles(len(xp))
    p[:, idx] = np.interp(x[:, idx], xp, fp)

return p
```

(same comment for `ppf` below)

Reply:
I didn't do this just because i kind of like assigning p/x to the scipy output so it's clear its equivalent to the empirical function.